### PR TITLE
Turbulent Scattering

### PR DESCRIPTION
**Pull request by profxj**
Methods to calculate turbulent scattering from the ISM.

Follows the formalism of Macquart&Koay13: http://adsabs.harvard.edu/abs/2013ApJ...776..125M
as expanded by Prochaska&Neeleman 2017.

docs and tests

### CODE REVIEW

File: frb/turb_scattering.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        ----------
+        ne : Quantity, optional
+          Electron density 
+        l0 : Quantity
+          Inner scale
+        L0 : Quantity, optional
+          Outer scale
+        SM : Quantity, optional
+          Generally calculated but can be input
+        zL : float, optional
```
Comment (mneeleman) [inline frb/turb_scattering.py]: why does z_L get treated differently? Isn't this just as much a property as L0, lo, and ne of the turbulence object. In other words, not setting it does not give any errors, yet it is likely not giving the correct value (unless the object you consider is at z=0).

File: docs/tscatt.rst [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+(or frequency).  Here is an example::
+
+    theta = turb.angular_broadening(lobs, zsource) # Returned in arcsec
+
+
+Intervening Galaxies
+--------------------
+
+One may generate a set of random draws for temporal smearing
+due to intervening galaxies towards a source with given
```
[the remark this answers is not in the document]
Comment (profxj) [reply docs/tscatt.rst]: done

File: frb/dlas.py [hunk 3]
```diff
@@ -160,6 +161,104 @@ def monte_DM(zeval, model='atan', nrand=100, verbose=False):
...
+    Parameters
+    ----------
+    zeval : ndarray
+      Array of redshifts for evaluation
+    nrand : int, optional
+      Number of samples on NHI
+    avg_ne : float, optional
+      Average log10 electron density / cm**3
+    sigma_ne : float, optional
+      Error in log10 ne
```
[the remark this answers is not in the document]
Comment (profxj) [reply frb/dlas.py]: yup and yup

File: frb/dlas.py [hunk 4]
```diff
@@ -160,6 +161,104 @@ def monte_DM(zeval, model='atan', nrand=100, verbose=False):
...
+    nHI : float, optional
+      Fiducial value for n_HI;  used for DL value
+    lobs : Quantity
+      Wavelength for analysis
+    turb : Turbulence object
+      Usually defined internally
```
[the remark this answers is not in the document]
Comment (profxj) [reply frb/dlas.py]: good call

File: frb/dlas.py [hunk 5]
```diff
@@ -160,6 +161,104 @@ def monte_DM(zeval, model='atan', nrand=100, verbose=False):
...
+
+    # Setup z
+    zvals = np.linspace(0., 7., 10000)
+    nz_s = _dla_nz(zvals)
+    nz_s[0] = 0.
+
+    # Turbulence
+    f_ne = 1e-3/u.cm**3  # Scale off of this
+    if turb is None:
+        turb = _init_dla_turb(ne=f_ne)
```
[the remark this answers is not in the document]
Comment (profxj) [reply frb/dlas.py]: good call;  done

File: frb/dlas.py [hunk 6]
```diff
@@ -203,6 +302,63 @@ def _atan_lz(zval, param=None):
     lz = param['A'] + param['B'] * np.arctan(zval-param['C'])
     return lz
 
+
+def _dla_nz(zarr, mlz=None, model='atan', analy_pth=None):
```
[the remark this answers is not in the document]
Comment (profxj) [reply frb/dlas.py]: nothing;  removed

File: frb/turb_scattering.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+# Constants
+const_re = const.alpha**2 * const.a0
+
+
+class Turbulence(object):
+    """ Class for turbulence calculations in a plasma
+    Primarily used for scattering calculations
+    """
+    def __init__(self, ne, l0, L0, zL=0., beta=11./3, SM=None, verbose=True):
```
[the remark this answers is not in the document]
Comment (profxj) [reply frb/turb_scattering.py]: Good suggestion.
Done.

File: frb/turb_scattering.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+        """
+        Parameters
+        ----------
+        ne : Quantity, optional
```
[the remark this answers is not in the document]
Comment (profxj) [reply frb/turb_scattering.py]: right you are

File: frb/turb_scattering.py (hunk 1 above)
Comment (profxj) [reply frb/turb_scattering.py]: right, it simply defaults to 0

File: frb/turb_scattering.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+        Equation 29 from Macquarty & Koay 2013
+        Assumes Kolmogorov
+        
+        Returns
+        -------
+        CN2 : Quantity
+        """
+        # Check
+        if (self.ne is None) or (self.L0 is None):
+            raise ValueError("Must specify ne and L0")
```
[the remark this answers is not in the document]
Comment (profxj) [reply frb/turb_scattering.py]: agreed.  Dang, I was sloppy!

File: frb/turb_scattering.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        # Simple expression
+        CN2 = 1.8e-3 * (self.ne/(1e-2*u.cm**(-3)))**2 * (self.L0/(0.001*u.pc))**(-2/3)
+        return CN2 * u.m**(-20/3.)
```
[the remark this answers is not in the document]
Comment (profxj) [reply frb/turb_scattering.py]: odd.  But good suggestion

File: frb/turb_scattering.py [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+        Parameters 
+        ----------
+        DL : Quantity
+          Thickness of the object
+
+        Returns
+        -------
+
+        """
+        self.SM = self.CN2_gal * DL
```
[the remark this answers is not in the document]
Comment (profxj) [reply frb/turb_scattering.py]: added .decompose()

File: frb/turb_scattering.py [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+        # Check
+        if self.SM is None:
+            raise IOError("Need to set SM first!")
+        # Useful expression
+        C = (np.pi*const_re**2 * lobs**2)/(1+self.zL)**2
+        # Is rdiff < l0?
+        r1 = 1. / np.sqrt(C * self.SM * (self.l0**(self.beta-4.) * (self.beta/4.) *
+          gamma(-self.beta/2.)))
+        # Is rdiff >> l0?
+        r2 = np.power(2**(2-self.beta) * C * self.SM * gamma(-self.beta/2.) /
```
[the remark this answers is not in the document]
Comment (profxj) [reply frb/turb_scattering.py]: egads!  Right you are.

I don't think it affects the FRB-DLA paper (we were in the other regime)

File: frb/turb_scattering.py [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+          gamma(self.beta/2.), 1./(2-self.beta))
+        # Query
+        if r1 < self.l0:
+            if self.verbose:
+                print('In the regime rdiff < l_0')
+            self.rdiff = r1.to('m')
+            self.regime = 1  # rdiff < l0
+        elif r2 > 10*self.l0:
```
[the remark this answers is not in the document]
Comment (profxj) [reply frb/turb_scattering.py]: added, with a warning.

**Comment by profxj** [comment]
I think we are ready @mneeleman

**Comment by profxj** [comment]
merging

**Pull request merged**